File: apps/server/openhive/auth/flows.py
"""In-memory flow registry tracking in-flight OAuth dances.

Each pending connection gets a short-lived FlowState keyed by a random flow_id returned
to the client. The state holds enough to resume the token exchange when the callback
arrives (PKCE verifier, expected state, provider, redirect URI for auth-code; device
code payload for device-flow). Flow records expire 5 minutes after creation.
"""

from __future__ import annotations

import asyncio
import secrets
import time
from dataclasses import dataclass, field
from typing import Any, Literal


FlowKind = Literal["auth_code", "device_code"]
FlowStatus = Literal["pending", "connected", "error", "expired"]


@dataclass
class FlowState:
    flow_id: str
    provider_id: str
    kind: FlowKind
    status: FlowStatus = "pending"
    error: str | None = None
    created_at: float = field(default_factory=time.time)
    # auth_code specific
    code_verifier: str | None = None
    expected_state: str | None = None
    redirect_uri: str | None = None
    # device_code specific — raw response retained for polling
    device_code: str | None = None
    user_code: str | None = None
    verification_uri: str | None = None
    verification_uri_complete: str | None = None
    device_interval: int | None = None
    device_expires_at: float | None = None
    # result
    account_label: str | None = None


_flows: dict[str, FlowState] = {}
_lock = asyncio.Lock()

FLOW_TTL_SECONDS = 300  # 5 min


def new_flow_id() -> str:
    return secrets.token_urlsafe(16)


async def create(provider_id: str, kind: FlowKind, **fields: Any) -> FlowState:
    async with _lock:
        state = FlowState(flow_id=new_flow_id(), provider_id=provider_id, kind=kind, **fields)
        _flows[state.flow_id] = state
        _gc_locked()
        return state


async def get(flow_id: str) -> FlowState | None:
    async with _lock:
        _gc_locked()
        return _flows.get(flow_id)


async def update(flow_id: str, **fields: Any) -> FlowState | None:
    async with _lock:
        state = _flows.get(flow_id)
        if not state:
            return None
        for k, v in fields.items():
            setattr(state, k, v)
        return state


async def remove(flow_id: str) -> None:
    async with _lock:
        _flows.pop(flow_id, None)


def _gc_locked() -> None:
    now = time.time()
    stale = [fid for fid, f in _flows.items() if now - f.created_at > FLOW_TTL_SECONDS]
    for fid in stale:
        _flows.pop(fid, None)

File: apps/web/lib/server/auth/flows.ts
/**
 * In-memory OAuth flow registry. Ports apps/server/openhive/auth/flows.py.
 *
 * Each pending connection gets a short-lived FlowState keyed by a random
 * flow_id returned to the client. State expires 5 minutes after creation.
 *
 * Cached on globalThis so Next.js HMR doesn't lose in-flight flows.
 */

import crypto from 'node:crypto'

type FlowKind = 'auth_code' | 'device_code'
type FlowStatus = 'pending' | 'connected' | 'error' | 'expired'

export interface FlowState {
  flow_id: string
  provider_id: string
  kind: FlowKind
  status: FlowStatus
  error: string | null
  created_at: number
  // auth_code
  code_verifier?: string | null
  expected_state?: string | null
  redirect_uri?: string | null
  // device_code
  device_code?: string | null
  user_code?: string | null
  verification_uri?: string | null
  verification_uri_complete?: string | null
  device_interval?: number | null
  device_expires_at?: number | null
  // result
  account_label?: string | null
}

interface FlowsCache {
  map: Map<string, FlowState>
}

const globalForFlows = globalThis as unknown as {
  __openhive_flows?: FlowsCache
}

function store(): Map<string, FlowState> {
  if (!globalForFlows.__openhive_flows) {
    globalForFlows.__openhive_flows = { map: new Map() }
  }
  return globalForFlows.__openhive_flows.map
}

const TTL_MS = 300_000

function gc(): void {
  const now = Date.now()
  const s = store()
  for (const [id, state] of s) {
    if (now - state.created_at > TTL_MS) s.delete(id)
  }
}

function newFlowId(): string {
  return crypto.randomBytes(16).toString('base64url')
}

export function createFlow(
  providerId: string,
  kind: FlowKind,
  fields: Partial<FlowState> = {},
): FlowState {
  gc()
  const state: FlowState = {
    flow_id: newFlowId(),
    provider_id: providerId,
    kind,
    status: 'pending',
    error: null,
    created_at: Date.now(),
    ...fields,
  }
  store().set(state.flow_id, state)
  return state
}

export function getFlow(flowId: string): FlowState | null {
  gc()
  return store().get(flowId) ?? null
}

export function updateFlow(
  flowId: string,
  fields: Partial<FlowState>,
): FlowState | null {
  const state = store().get(flowId)
  if (!state) return null
  Object.assign(state, fields)
  return state
}

function removeFlow(flowId: string): void {
  store().delete(flowId)
}
